js_modules/dagit/packages/core/src/ui/BaseButton.tsx

```diff
@@ -0,0 +1,147 @@
+import * as React from 'react';
+import styled, {css} from 'styled-components/macro';
+
+import {ColorsWIP} from './Colors';
+import {IconWrapper} from './Icon';
+
+interface Props extends React.ComponentPropsWithRef<'button'> {
+  icon?: React.ReactNode;
+  label?: React.ReactNode;
+  loading?: boolean;
+  rightIcon?: React.ReactNode;
+  fillColor?: string;
+  stroke?: boolean;
+  textColor?: string;
+}
+
+export const BaseButton = React.forwardRef(
+  (props: Props, ref: React.ForwardedRef<HTMLButtonElement>) => {
+    const {
+      fillColor = ColorsWIP.White,
+      disabled,
+      icon,
+      label,
+      loading,
+      rightIcon,
+      textColor = ColorsWIP.Dark,
+      stroke = true,
+      ...rest
+    } = props;
+
+    return (
+      <StyledButton
+        {...rest}
+        disabled={disabled || loading}
+        $fillColor={fillColor}
+        $stroke={stroke}
+        $textColor={textColor}
+        ref={ref}
+      >
+        {icon || null}
+        {label ? <span>{label}</span> : null}
+        {rightIcon || null}
+      </StyledButton>
+    );
+  },
+);
+
+interface StyledButtonProps {
+  $fillColor: string;
+  $stroke: boolean;
+  $textColor: string;
+}
+
+const backgroundColorCSS = (props: StyledButtonProps) => {
+  const {$fillColor} = props;
+  if ($fillColor) {
+    return $fillColor;
+  }
+  return 'transparent';
+};
+
+const DEFAULT_STROKE = css`
+  box-shadow: rgba(0, 0, 0, 0) 0px 1px 1px 0px, ${ColorsWIP.Gray300} 0px 0px 0px 1px;
+`;
+const NO_STROKE = css`
+  box-shadow: rgba(0, 0, 0, 0) 0px 1px 1px 0px, rgba(0, 0, 0, 0) 0px 0px 0px 1px;
+`;
+const DEFAULT_STROKE_PLUS_HOVER = css`
+  box-shadow: rgb(0, 0, 0, 0) 0px 1px 1px 0px, ${ColorsWIP.Gray400} 0px 0px 0px 1px,
+    rgba(0, 0, 0, 0.12) 0px 2px 12px 0px;
+`;
+const NO_STROKE_PLUS_HOVER = css`
+  box-shadow: rgba(0, 0, 0, 0) 0px 1px 1px 0px, rgba(0, 0, 0, 0) 0px 0px 0px 1px,
+    rgba(0, 0, 0, 0.12) 0px 2px 12px 0px;
+`;
+
+const boxShadowCSS = (props: StyledButtonProps) => {
```

Comment:
Are these factored out of the template string below so that they aren't recomputed? All else being equal I probably would have inlined all of this into the StyleButton declaration, but maybe I shouldn't be doing that 😅 . It'd be a bit easier to "see" what the result looked like if it was just:

 `background-color: ${({$fillColor}) => $fillColor || 'transparent'}`